cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageTable.java

```diff
@@ -0,0 +1,484 @@
+    // Similarly once the time is inverted, original start-time becomes the end time of the scan. Since we want to
+    // return the records corresponding to this end time (inverted start-time), we add 1.
+    long invertedTime = invertTime(time);
+    invertedTime = invertedTime == Long.MAX_VALUE ? invertedTime : invertedTime + 1;
+    fields.add(Fields.longField(StoreDefinition.FieldLineageStore.START_TIME_FIELD, invertedTime));
+    return fields;
+  }
+
+  private List<Field<?>> getSummaryPrefix(long checksum, String direction, EndPoint endPoint) {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.longField(StoreDefinition.FieldLineageStore.CHECKSUM_FIELD, checksum));
+    fields.add(Fields.stringField(StoreDefinition.FieldLineageStore.DIRECTION_FIELD, direction));
+    addEndPoint(fields, endPoint);
+    return fields;
+  }
+
+  private List<Field<?>> getOperationReferenceRowKey(String direction, EndPoint endPoint, ProgramRunId programRunId) {
+    long invertedStartTime = getInvertedStartTime(programRunId);
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.stringField(StoreDefinition.FieldLineageStore.DIRECTION_FIELD, direction));
+    addEndPoint(fields, endPoint);
+    fields.add(Fields.longField(StoreDefinition.FieldLineageStore.START_TIME_FIELD, invertedStartTime));
+    return fields;
+  }
+
+  private List<Field<?>> getDestinationKeys(long checksum, EndPoint endPoint) {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.longField(StoreDefinition.FieldLineageStore.CHECKSUM_FIELD, checksum));
+    addEndPoint(fields, endPoint);
+    return fields;
+  }
+
+  private long invertTime(long time) {
+    return Long.MAX_VALUE - time;
+  }
+
+  private long getInvertedStartTime(ProgramRunId run) {
+    return invertTime(RunIds.getTime(RunIds.fromString(run.getEntityName()), TimeUnit.MILLISECONDS));
+  }
+
+  private void addEndPoint(List<Field<?>> fields, EndPoint endPoint) {
+    fields.add(Fields.stringField(StoreDefinition.FieldLineageStore.ENDPOINT_NAMESPACE_FIELD, endPoint.getNamespace()));
+    fields.add(Fields.stringField(StoreDefinition.FieldLineageStore.ENDPOINT_NAME_FIELD, endPoint.getName()));
+  }
+
+  private List<Field<?>> getSummaryKey(long checksum, String direction, EndPointField endPointField) {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.longField(StoreDefinition.FieldLineageStore.CHECKSUM_FIELD, checksum));
+    fields.add(Fields.stringField(StoreDefinition.FieldLineageStore.DIRECTION_FIELD, direction));
+    addEndPoint(fields, endPointField.getEndPoint());
+    fields.add(Fields.stringField(StoreDefinition.FieldLineageStore.ENDPOINT_FIELD, endPointField.getField()));
+    return fields;
+  }
+
+  private List<Field<?>> getOperationsKey(long checksum) {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.longField(StoreDefinition.FieldLineageStore.CHECKSUM_FIELD, checksum));
+    return fields;
```

Comment:
I think you can use Collections.singletonList here.

Reply:
prefer to keep things mutable here since this is used to construct larger keys